docs/proposals/ssh_adapter.md

```diff
@@ -0,0 +1,194 @@
+# SSH protocol proposal
+
+We'd like to implement a protocol for Git LFS that uses SSH protocol
+exclusively, avoiding the need to use HTTPS altogether.  This will make
+deployment and use easier in a variety of situations where access to certain
+ports is limited.
+
+## What not to do
+
+There are several possible approaches that could be adopted.  SSH provides a
+native capability for the SFTP protocol, which can be used to transfer files.
+However, in order to implement this on the server side, each access (upload or
+download) must have an access control check instead of one at the beginning of
+the operation.  This would be inefficient in some server-side implementations,
+and nearly impossible to implement securely for implementations that use the
+system OpenSSH for implementation.
+
+## A more usable approach
+
+Git already has some places we can look for inspiration.  Its SSH protocol is
+based on the Git native protocol, which is based on the pkt-line scheme.
```

Comment:
I don't think that this is related to your proposal at all, but we do have a good implementation of pkt-line in package `git`, thanks to @larsxschneider's work during the Git 2.11-era push towards `filter.lfs.process`.

I guess all of that is to say: this makes implementing a pkt-line-based protocol less work for us, since we don't have to first implementation serialization over that protocol.

But, one thing we definitely _don't_ have is an implementation that will provide us convenient data structures over protocol v2 concepts, so I think that the work here lays there. Perhaps pkt-line and this proposed new implementation should reside in a separate repository?

Reply:
It's possible.  I don't know how many people are going to find pkt-line work in Go super useful; I expect all of the people who are going to be doing that work are implementers of Git-related things, either client or server side.  It may be useful to start work and then split things out if we find a logical unit that works on its own.